pkg/controller/service/service_controller_test.go

```diff
@@ -152,61 +195,205 @@ func TestCreateExternalLoadBalancer(t *testing.T) {
+					SelfLink:   testapi.Default.SelfLink("services", "basic-service1"),
+					Finalizers: []string{servicehelper.LoadBalancerCleanupFinalizer},
+				},
+				Spec: v1.ServiceSpec{
+					Ports: []v1.ServicePort{{
+						Port:     80,
+						Protocol: v1.ProtocolTCP,
+					}},
+					Type: v1.ServiceTypeLoadBalancer,
+				},
+			},
+			expectOp:             ensureLoadBalancer,
+			expectCreateAttempt:  true,
+			expectPatchStatus:    true,
+			expectPatchFinalizer: false,
 		},
 	}
 
-	for _, item := range table {
-		controller, cloud, client := newController()
-		key := fmt.Sprintf("%s/%s", item.service.Namespace, item.service.Name)
-		if _, err := client.CoreV1().Services(item.service.Namespace).Create(item.service); err != nil {
-			t.Errorf("Failed to prepare service %s for testing: %v", key, err)
-			continue
-		}
-		client.ClearActions()
-		err := controller.syncLoadBalancerIfNeeded(key, item.service)
-		if !item.expectErr && err != nil {
-			t.Errorf("unexpected error: %v", err)
-		} else if item.expectErr && err == nil {
-			t.Errorf("expected error creating %v, got nil", item.service)
-		}
-		actions := client.Actions()
-		if !item.expectCreateAttempt {
-			if len(cloud.Calls) > 0 {
-				t.Errorf("unexpected cloud provider calls: %v", cloud.Calls)
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceLoadBalancerFinalizer, tc.enableFeatureGate)()
+
+			controller, cloud, client := newController()
+			cloud.Exists = tc.lbExists
+			key := fmt.Sprintf("%s/%s", tc.service.Namespace, tc.service.Name)
+			if _, err := client.CoreV1().Services(tc.service.Namespace).Create(tc.service); err != nil {
+				t.Fatalf("Failed to prepare service %s for testing: %v", key, err)
+			}
+			client.ClearActions()
+
+			op, err := controller.syncLoadBalancerIfNeeded(tc.service, key)
+			if err != nil {
+				t.Errorf("Got error: %v, want nil", err)
 			}
-			if len(actions) > 0 {
-				t.Errorf("unexpected client actions: %v", actions)
+			if op != tc.expectOp {
+				t.Errorf("Got operation %v, want %v", op, tc.expectOp)
 			}
-		} else {
-			var balancer *fakecloud.Balancer
-			for k := range cloud.Balancers {
+			// Capture actions from test so it won't be messed up.
+			actions := client.Actions()
+
+			if !tc.expectCreateAttempt && !tc.expectDeleteAttempt {
+				if len(cloud.Calls) > 0 {
+					t.Errorf("Unexpected cloud provider calls: %v", cloud.Calls)
+				}
+				if len(actions) > 0 {
+					t.Errorf("Unexpected client actions: %v", actions)
+				}
+				return
+			}
+
+			if tc.expectCreateAttempt {
+				createCallFound := false
+				for _, call := range cloud.Calls {
+					if call == "create" {
+						createCallFound = true
+					}
+				}
+				if !createCallFound {
+					t.Errorf("Got no create call for load balancer, expected one")
+				}
+				var balancer *fakecloud.Balancer
+				for k := range cloud.Balancers {
```

Comment:
Probably for a follow-up since this was here before your PR, but this pattern seems awkward, can just be:
```go
if len(cloud.Balancers) != 1 {
    t.Errorf("Expected exactly 1 load balancer, got %v, cloud.Balancers)
} else {
    b := cloud.Balancers[0]
    balancer = &b
}
```

Reply:
Agreed, I wanted to clean this up but decided to do it later. Added a TDO to follow up on this.